src/components/orator/js/Orator.js

```diff
@@ -0,0 +1,571 @@
+                        blockIndex = lastParsed.blockIndex + lastParsed.word.length;
+                    }
+                }
+            });
+        }
+
+        return parsed;
+    };
+
+    /**
+     * Combines the parsed text into a String.
+     *
+     * @param {Array} parsed - An array of parsed data points
+     *
+     * @return {String} - The parsed text combined into a String.
+     */
+    fluid.orator.domReader.parsedToString = function (parsed) {
+        var words = fluid.transform(parsed, function (block) {
+            return block.word;
+        });
+
+        return words.join("");
+    };
+
+    /**
+     * Parses the DOM element into data points to use for highlighting the text, and queues the text into the self
+     * voicing engine. The parsed data points are added to the component's `parseQueue`
+     *
+     * @param {Component} that - the component
+     * @param {String|jQuery|DomElement} elm - The DOM node to read
+     */
+    fluid.orator.domReader.readFromDOM = function (that, elm) {
+        elm = $(elm);
+
+        // only execute if there are nodes to read from
+        if (elm.length) {
+            var parsedFromElm = fluid.orator.domReader.parse(elm[0]);
+            that.parseQueue = parsedFromElm;
+            that.events.onReadFromDOM.fire(parsedFromElm);
+            that.tts.queueSpeech(fluid.orator.domReader.parsedToString(parsedFromElm));
+        }
+    };
+
+    /**
+     * Returns the index of the closest data point from the parseQueue based on the boundary provided.
+     *
+     * @param {Array} parseQueue - An array data points generated from parsing a DOM structure
+     * @param {Integer} currentIndex - The index into the paraseQueue to start searching from. The currentIndex will be
+     *                                constrained to the bounds of the parseQueue.
+     * @param {Integer} boundary - The boundary value used to compare against the blockIndex of the parsed data points.
+     *
+     * @return {Integer|undefined} - Will return the index of the closest data point in the parseQueue. If the boundary
+     *                               cannot be located within the parseQueue, `undefined` is returned.
+     */
+    fluid.orator.domReader.getClosestIndex = function (parseQueue, currentIndex, boundary) {
+        var maxIdx  = Math.max(parseQueue.length - 1, 0);
+        currentIndex = Math.max(Math.min(currentIndex, maxIdx), 0);
+
+        var nextIdx = currentIndex + 1;
+        var prevIdx = currentIndex - 1;
+
+        var currentBlockIndex = parseQueue[currentIndex].blockIndex;
+        var maxBoundary = parseQueue[maxIdx].blockIndex + parseQueue[maxIdx].word.length;
+
+
+        if (!fluid.isValue(boundary) || boundary < 0 || boundary > maxBoundary ) {
+            return undefined;
+        }
+
+        if (currentBlockIndex > boundary) {
+            return fluid.orator.domReader.getClosestIndex(parseQueue, prevIdx, boundary);
```

Comment:
I still don't see any reason for this function to be recursive, least of all doubly recursive, and would be much clearer if it was rewritten so that all the initial setup happens just once

Reply:
Explore re-writing this as a while loop with break conditions.